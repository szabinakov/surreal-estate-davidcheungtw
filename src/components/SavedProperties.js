import React, { useState, useEffect } from "react";
import axios from "axios";
import PropTypes from "prop-types";
import Alert from "./Alert";
import PropertyCard from "./PropertyCard";
import getSavedProperties from "../requests/getSavedProperties";
import "../styles/saved-properties.css";

const SavedProperties = ({ userID }) => {
  const [properties, setProperties] = useState([]);
  const [alert, setAlert] = useState({ message: "", isSuccess: true });

  useEffect(() => {
    getSavedProperties(setProperties, setAlert);
  }, []);

  const removeProperty = (id) => {
    axios
      .delete(`http://localhost:4000/api/v1/Favourite/${id}`)
      .then(() => {
        setAlert({
          message: "Saved property removed.",
          isSuccess: true,
        });
      })
      .catch(() => {
        setAlert({
          message: "Server error. Please try again later.",
          isSuccess: false,
        });
      });
  };

  if (!userID) {
    return <div className="saved-properties">Please login first</div>;
  }

  return (
    <div className="saved-properties">
      {alert.message && (
        <Alert message={alert.message} isSuccess={alert.isSuccess} />
      )}
      <h5> User ID : {userID}</h5>
      {!alert.message && (
        <div className="property-card-list">
          {properties.map((property) => (
            <div key={property._id} className="item">
              <PropertyCard
                key={property._id}
                {...property.propertyListing}
                userID={userID}
                onSaveProperty={() => {}}
                removeProperty={removeProperty}
                removeId={property._id}
              />
            </div>
          ))}
        </div>
      )}
    </div>
  );
};

export default SavedProperties;

SavedProperties.propTypes = {
  userID: PropTypes.string.isRequired,
};
