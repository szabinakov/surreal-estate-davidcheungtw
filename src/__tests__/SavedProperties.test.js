import React from "react";
import { BrowserRouter as Router } from "react-router-dom";
import { render, screen, waitFor } from "@testing-library/react";
import renderer from "react-test-renderer";
import axios from "axios";
import SavedProperties from "../components/SavedProperties";

describe("SavedProperties", () => {
  const validProps = {
    userID: "111222333444",
  };
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("Renders as expected", () => {
    const rendered = renderer.create(
      <Router>
        <SavedProperties userID={validProps.userID} />
      </Router>
    );

    expect(rendered).toMatchSnapshot();
  });

  it("Renders success post request to Surreal Estate API", async () => {
    const mAxiosResponse = {
      data: [
        {
          _id: "01010101",
          propertyListing: {
            _id: "643a793b396cf20f05fb487c",
            title: "title001",
            type: "Flat",
            bedrooms: "1",
            bathrooms: "1",
            price: "1000",
            city: "Manchester",
            email: "joe@mcr.codes",
            __v: 0,
          },
        },
        {
          _id: "02020202",
          propertyListing: {
            _id: "643a795a396cf24c7cfb487e",
            title: "title002",
            type: "Semi-detached",
            bedrooms: "2",
            bathrooms: "2",
            price: "2000",
            city: "Leeds",
            email: "joe@mcr.codes",
            __v: 0,
          },
        },
        {
          _id: "03030303",
          propertyListing: {
            _id: "643a79d1396cf25d8dfb4883",
            title: "title003",
            type: "Detached",
            bedrooms: "3",
            bathrooms: "3",
            price: "3000",
            city: "Liverpool",
            email: "joe@mcr.codes",
            __v: 0,
          },
        },
      ],
    };
    jest.spyOn(axios, "get").mockResolvedValue(mAxiosResponse);
    render(
      <Router>
        <SavedProperties userID={validProps.userID} />
      </Router>
    );

    await waitFor(() => {
      expect(screen.getByText("Flat - Manchester")).toBeInTheDocument();
      expect(screen.getByText("Semi-detached - Leeds")).toBeInTheDocument();
      expect(screen.getByText("Detached - Liverpool")).toBeInTheDocument();
    });
  });
});
