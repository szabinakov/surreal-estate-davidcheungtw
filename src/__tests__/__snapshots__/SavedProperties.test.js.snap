// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SavedProperties Renders as expected 1`] = `
<div
  className="saved-properties"
>
  <h5>
     User ID : 
    111222333444
  </h5>
  <div
    className="property-card-list"
  />
</div>
`;
